src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -86,86 +139,138 @@ namespace NovelRT::Graphics
+                                                                                         size_t dataLength,
+                                                                                         size_t alignment)
+    {
+        size_t sizeToStage = dataTypeSize * dataLength;
+        auto currentContext = _graphicsDevice->GetCurrentContext();
+
+        std::shared_ptr<GraphicsBuffer> stagingBuffer =
+            GetStagingBufferWithProperSizeHandling(sizeToStage, currentContext);
+
+        std::shared_ptr<GraphicsBuffer> indexBuffer =
+            GetOrCreateGraphicsBufferForAllocationSize(sizeToStage, GraphicsBufferKind::Index);
+        auto indexRegion = indexBuffer->Allocate(sizeToStage, alignment);
+        auto writeArea = stagingBuffer->Map<uint8_t>(indexRegion);
+#ifdef WIN32
+        memcpy_s(writeArea, sizeToStage, data, sizeToStage);
+#else
+        memcpy(writeArea, data, sizeToStage);
+#endif
+        stagingBuffer->UnmapAndWrite(indexRegion);
+        currentContext->Copy(indexBuffer, stagingBuffer);
+
+        return indexRegion;
+    }
+
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetStagingBufferWithProperSizeHandling(
         size_t sizeToStage,
-        std::shared_ptr<GraphicsContext>& currentContext)
+        std::shared_ptr<GraphicsContext>& /*currentContext*/)
     {
-        std::shared_ptr<GraphicsBuffer> stagingBuffer = _stagingBuffer.getActual();
+        _stagingBufferSize = _stagingBufferSize < sizeToStage ? sizeToStage : _stagingBufferSize;
+
+        std::shared_ptr<GraphicsBuffer> chosenBuffer = nullptr;
+
+        for (auto&& buffer : _stagingBuffers[_contextIndex])
+        {
+            if (buffer->GetLargestFreeRegionSize() >= sizeToStage)
+            {
+                chosenBuffer = buffer;
+                break;
+            }
+        }
 
-        if (stagingBuffer->GetLargestFreeRegionSize() < sizeToStage)
+        if (chosenBuffer == nullptr)
         {
-            _stagingBufferSize = (_stagingBufferSize * 2) + sizeToStage;
-            std::shared_ptr<GraphicsBuffer> newStagingBuffer = CreateStagingBuffer();
-            currentContext->Copy(newStagingBuffer, stagingBuffer);
-            currentContext->EndFrame();
-            _graphicsDevice->Signal(currentContext->GetFence());
-            _graphicsDevice->WaitForIdle();
-            _stagingBuffer.reset(newStagingBuffer);
-            stagingBuffer = _stagingBuffer.getActual();
-            currentContext->BeginFrame();
+            chosenBuffer = CreateStagingBuffer();
         }
-        return stagingBuffer;
+
+        return chosenBuffer;
     }
 
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::CreateStagingBuffer()
     {
-        return _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
+        auto bufferToReturn = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
```

Comment:
Might be better to just place this entire thing in the `emplace_back` so that it can move/whatever as appropriate.